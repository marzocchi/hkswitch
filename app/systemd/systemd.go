package systemd

import (
	"bytes"
	_ "embed"
	"fmt"
	"io"
	"text/template"
)

//go:embed systemd.service
var systemdTemplateText string

//go:embed launchd.plist
var launchdTemplateText string

type EnvVar struct {
	Name  string
	Value string
}

type Unit struct {
	WorkingDir  string
	Description string
	CommandLine string
	Command     []string
	Home        string
	Env         []EnvVar
}

func (c *Unit) Write(w io.Writer, initType string) error {
	tpl, err := c.getTemplate(initType)
	if err != nil {
		return err
	}

	buf := &bytes.Buffer{}

	if err := tpl.Execute(buf, c); err != nil {
		return err
	}

	if _, err := io.Copy(w, buf); err != nil {
		return err
	}

	return nil
}

func (c *Unit) getTemplate(initType string) (*template.Template, error) {
	var templateText string
	switch initType {
	case "systemd":
		templateText = systemdTemplateText
	case "launchd":
		templateText = launchdTemplateText
	default:
		return nil, fmt.Errorf("not supported: %q", initType)
	}

	tpl := template.New("init-template")
	tpl.Option("missingkey=error")

	t, err := tpl.Parse(templateText)
	if err != nil {
		return nil, err
	}

	return t, nil
}

