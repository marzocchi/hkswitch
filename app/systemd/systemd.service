[Unit]
Description={{ .Description }}
Requires=network.target local-fs.target
After=network.target local-fs.target

[Install]
WantedBy=multi-user.target

[Service]
Type=simple
User=root
WorkingDirectory={{ .WorkingDir }}
Restart=on-failure
ExecStart={{ .CommandLine }}
ExecStop=/bin/kill -INT $MAINPID
TimeoutStopSec=10
{{ range .Env }}Environment={{ .Name }}="{{ .Value }}"
{{ end }}
