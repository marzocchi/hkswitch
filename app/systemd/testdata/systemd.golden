[Unit]
Description=test
Requires=network.target local-fs.target
After=network.target local-fs.target

[Install]
WantedBy=multi-user.target

[Service]
Type=simple
User=root
WorkingDirectory=/var/run
Restart=on-failure
ExecStart=cmd arg
ExecStop=/bin/kill -INT $MAINPID
TimeoutStopSec=10
Environment=VARNAME1="VARVALUE1"
Environment=VARNAME2="VARVALUE2"
